test/Helper/Service/DoctypeFactoryTest.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/**
+ * @see       https://github.com/laminas/laminas-view for the canonical source repository
+ * @copyright https://github.com/laminas/laminas-view/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-view/blob/master/LICENSE.md New BSD License
+ */
+
+namespace LaminasTest\View\Helper\Service;
+
+use Interop\Container\ContainerInterface;
+use Laminas\ServiceManager\ServiceManager;
+use Laminas\View\Helper\Doctype;
+use Laminas\View\Helper\Service\DoctypeFactory;
+use PHPUnit\Framework\TestCase;
+
+final class DoctypeFactoryTest extends TestCase
+{
+    protected function setUp(): void
+    {
+        Doctype::unsetDoctypeRegistry();
+    }
+
+    public function testServiceIsCreatedOnInvocation(): void
+    {
+        $container = $this->createContainer();
```

Comment:
A mock object should be used here which is build by PHPUnit. The interface of the container is relevant here, not the concrete implementation like the service manager.
This saves us work in the future.

Thanks in advance!

Reply:
> A mock object should be used here which is build by PHPUnit. The interface of the container is relevant here, not the concrete implementation like the service manager.
> This saves us work in the future.

Really? I almost never use mocks because of the [headache they caused](https://thephp.cc/dates/2020/06/php-fwdays-special-edition/avoiding-mockageddon) me in the past and always use real stuff, if possible. If the ServiceManager or Container changes, the test _should_ break because the Factory won't work in a real application anymore.
But if Mocks are the Laminas guidelines, I'll of course change the code accordingly.

On a side note: Personally, I even think that the whole mocking stuff should be refactored out of PHPUnit, just like Prophecy.

Reply:
> I almost never use mocks because of the headache they caused…

1. The dependency is the container interface, not the service manager.
2. There is no _overmocking_ in this case.

```php
$container->has('config')
```

https://github.com/laminas/laminas-view/pull/59/files#diff-e6d323bcbed448c76f7d30b10217611dc0ffd9ab40ae27272dd9fe17d87da035R31

```php
$container->get('config')
```
https://github.com/laminas/laminas-view/pull/59/files#diff-e6d323bcbed448c76f7d30b10217611dc0ffd9ab40ae27272dd9fe17d87da035R34

> If the ServiceManager or Container changes, the test should break because the Factory won't work in a real application anymore.

Create a mock object for the interface of the container then it does not matter if the service manager or the container changes.

> Personally, I even think that the whole mocking stuff should be refactored out of PHPUnit, just like Prophecy.

This is the wrong issue tracker! 😉

Reply:
done! I changed the test to use a Mock object